### PR TITLE
feat(warehouse): add ValidateSQL to Provider interface

### PR DESCRIPTION
**Pull request by abacigil**
## Summary

`warehouse.Provider` gains a `ValidateSQL(ctx, sql) error` method
so callers that need to know "would this SQL compile?" can ask the
warehouse itself, in its own dialect, without hand-rolled regex /
heuristic SQL inspection in caller code.

## Implementation per provider

Each routes through the warehouse's native compile-only path:

| Provider | Path |
|---|---|
| BigQuery | `QueryConfig.DryRun = true` |
| PostgreSQL | `EXPLAIN <sql>` |
| Redshift | `EXPLAIN <sql>` |
| Snowflake | `EXPLAIN USING TEXT <sql>` |
| Databricks | `EXPLAIN <sql>` |
| SQL Server | `SET NOEXEC ON; <sql>; SET NOEXEC OFF;` |

**T-SQL note:** SET PARSEONLY does NOT gate same-batch
execution — an integration test confirmed an INSERT ran through
a `SET PARSEONLY ON; INSERT …; SET PARSEONLY OFF;` batch and the
row count went up. MSSQL uses `NOEXEC` instead, which is
session-scoped and applies to the current batch.

## Contract

- Empty / whitespace-only SQL is rejected at the boundary.
- The warehouse's own error message is propagated verbatim
  (suitable for echoing into an end-user message).
- Implementations must NOT execute the statement. A passing
  `ValidateSQL` on a write statement does not authorise
  execution — read-only enforcement remains `ValidateReadOnly`'s
  responsibility.
- The check is "compile, don't execute": for every provider this
  includes syntax + name resolution + plan compilation. A query
  against a missing table surfaces as an error.

## Tests

- **Unit tests** for every provider cover: empty-input boundary,
  the wire format the implementation hands to the client, error
  propagation.
- **Integration tests** behind the existing per-provider build
  tags exercise the real compile path:
  - Postgres: testcontainers (`make test-postgres`)
  - BigQuery: real (`make test-bigquery` with
    `INTEGRATION_TEST_BIGQUERY_*` env vars; verified against
    `demo_olist_ecommerce.customers`)
  - Redshift: real (`make test-redshift` against `default-workgroup`
    on us-east-1)
  - MSSQL: testcontainers (`make test-mssql`)
  - Snowflake / Databricks: unit-only — no test accounts
    available yet.

The MSSQL integration test specifically validates an INSERT
through `ValidateSQL` and asserts the row count is unchanged, so
a future regression away from NOEXEC fails loudly.

## In-tree implementer updates

Test fakes that previously implemented `warehouse.Provider`
get the new method as a no-op so the agent test suite still
builds:

- `libs/go-common/warehouse/registry_test.go::mockWarehouseProvider`
- `services/agent/internal/testutil/mocks.go::MockWarehouseProvider`
- `services/agent/internal/queryexec/query_executor_test.go::queryWrapper`
- `services/agent/internal/discovery/prompt_subs_test.go::fakeProvider`

## Docs

- `docs/concepts/providers.md` — interface table updated with
  the new method.
- `docs/guides/adding-warehouse-providers.md` — interface
  listing + method-details table updated.
- `CHANGELOG.md` — entry under `[Unreleased]` describing the new
  method, per-provider implementations, the MSSQL PARSEONLY
  finding, and the pre-1.0 break.

## Breaking change (pre-1.0)

Any out-of-tree `warehouse.Provider` implementation must add the
new method. No shims; the project is pre-1.0.

## Test plan

- [x] `make test-go` — green
- [x] `make lint-go` — 0 issues
- [x] `make test-postgres` (testcontainers) — green
- [x] `make test-bigquery` against `demo_olist_ecommerce.customers` — green
- [x] `make test-redshift` against `default-workgroup` — green
- [x] `make test-mssql` (testcontainers) — green
- [ ] CI on this PR with `run-integration-tests` label

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: providers/warehouse/mssql/client.go [hunk 1]
```diff
@@ -9,6 +9,9 @@ import (
 // The real implementation is *sql.DB opened via the microsoft/go-mssqldb driver.
 type msClient interface {
 	QueryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error)
+	// ExecContext runs a statement that doesn't return rows (used
+	// by ValidateSQL's SET PARSEONLY ON/OFF batch).
```
[the remark this answers is not in the document]
Comment (abacigil) [reply providers/warehouse/mssql/client.go]: Addressed in e7c9d67. Updated the ExecContext comment to reference SET NOEXEC ON/OFF (the actual mechanism).

File: providers/warehouse/mssql/integration_test.go [hunk 2]
```diff
@@ -922,3 +922,105 @@ func assertGoType[T any](t *testing.T, row map[string]interface{}, col string) {
...
+		if err := p.ValidateSQL(context.Background(), sql); err == nil {
+			t.Errorf("ValidateSQL accepted whitespace-only input %q", sql)
+		}
+	}
+}
+
+func TestIntegration_ValidateSQL_DoesNotExecuteSideEffects(t *testing.T) {
+	// PARSEONLY checks syntax without compiling or executing, so an
+	// INSERT must NOT modify the table. Confirm by counting before
+	// and after a ValidateSQL call against an INSERT statement.
```
[the remark this answers is not in the document]
Comment (abacigil) [reply providers/warehouse/mssql/integration_test.go]: Addressed in e7c9d67. Test docstring now describes NOEXEC and explicitly notes this test is the regression pin against reverting to PARSEONLY.

File: providers/warehouse/mssql/integration_test.go [hunk 3]
```diff
@@ -922,3 +922,105 @@ func assertGoType[T any](t *testing.T, row map[string]interface{}, col string) {
...
+		t.Fatalf("count before: %v", err)
+	}
+	if len(beforeRes.Rows) != 1 {
+		t.Fatalf("count before: expected 1 row, got %d", len(beforeRes.Rows))
+	}
+	before := beforeRes.Rows[0]["c"]
+
+	// PARSEONLY validates the syntax of an INSERT and returns nil.
+	// Read-only enforcement is ValidateReadOnly's concern.
+	_ = p.ValidateSQL(ctx, "INSERT INTO all_types (col_nvarchar) VALUES (N'side effect')")
```
[the remark this answers is not in the document]
Comment (abacigil) [reply providers/warehouse/mssql/integration_test.go]: Addressed in e7c9d67. Inline comment now describes NOEXEC compile-only behaviour.

File: docs/guides/adding-warehouse-providers.md [hunk 4]
```diff
@@ -38,6 +39,7 @@ type Provider interface {
 | `QuoteRef` | Return a dialect-correct fully-qualified identifier | Quote each part with the dialect's native delimiter and join with dots: BigQuery / Databricks use backticks, PostgreSQL / Redshift / Snowflake use double quotes, SQL Server uses square brackets. Delegate to the colocated helper `warehouse.QuotePartsWith(open, close, parts)`. Used by the orchestrator to render `{{REF:table}}` placeholders in prompts. |
 | `SQLFixPrompt` | Return warehouse-specific SQL fix prompt | Instructions for the AI to fix SQL errors |
 | `ValidateReadOnly` | Verify read access works | Run a simple query to confirm connectivity |
+| `ValidateSQL` | Compile a SQL statement without executing it | Route through the warehouse's native compile-only path: `EXPLAIN <sql>` on Postgres-family / Databricks / Redshift, `EXPLAIN USING TEXT <sql>` on Snowflake, `QueryConfig.DryRun = true` on BigQuery, `SET NOEXEC ON; <sql>; SET NOEXEC OFF;` on SQL Server. Reject empty / whitespace SQL at the boundary and propagate the warehouse's own error verbatim. Must NOT execute the statement — read-only enforcement remains `ValidateReadOnly`'s concern. |
```
[the remark this answers is not in the document]
Comment (abacigil) [reply docs/guides/adding-warehouse-providers.md]: Addressed in e7c9d67. Reworded the docs: implementations wrap with `<provider>: validate SQL: %w` so the warehouse's underlying message is preserved alongside the routing context. Picked the doc-side fix rather than the implementation-side one — the wrapping is intentional so callers can tell which provider returned the error without parsing the message.

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
...e dialect-owner decides whether the statement parses, resolves names, and compiles: BigQuery uses `QueryConfig.DryRun = true`; PostgreSQL, Redshift, and Databricks use `EXPLAIN <sql>`; Snowflake uses `EXPLAIN USING TEXT <sql>`; SQL Server uses `SET NOEXEC ON; <sql>; SET NOEXEC OFF;` (PARSEONLY only applies to subsequent batches in T-SQL, so it cannot gate a same-batch INSERT — confirmed by integration test). All six providers reject empty / whitespace-only SQL at the boundary, propagate the warehouse's own error message verbatim, and explicitly do not execute the statement. Read-only enforcement remains `ValidateReadOnly`'s responsibility — passing `ValidateSQL` on a write statement does not authorise execution. Unit tests cover the boundary check, the wire-format expectations, and error propagation for every provider; integration tests behind the existing per-provider build tags exercise the real compile path on Postgres (testcontainers), BigQuery (real), Redshift (real), and SQL Server (testcontainers). Snowflake and Databricks ship unit-only coverage until test accounts are available. **Pre-1.0 break:** any out-of-tree `warehouse.Provider` implementation must add the new method.
```
[the remark this answers is not in the document]
Comment (abacigil) [reply CHANGELOG.md]: Addressed in e7c9d67. Same fix as the docs comment: reworded the CHANGELOG to describe the wrapped-error shape (`<provider>: validate SQL: %w`) instead of claiming verbatim propagation.

**Comment by abacigil** [comment]
## Self-review

### What changed since the initial commit

Single follow-up commit (e7c9d67) addressing five Copilot review findings — all documentation alignment, no behaviour change:

- Stale `SET PARSEONLY ON/OFF` references in MSSQL `client.go` / `integration_test.go` (left over from the initial pre-NOEXEC implementation) updated to NOEXEC.
- "propagate verbatim" claim in `docs/guides/adding-warehouse-providers.md` and `CHANGELOG.md` softened — implementations wrap the warehouse error with a `<provider>: validate SQL: %w` prefix so the underlying message is preserved alongside provider context. The wrapping is intentional (so callers can tell which provider returned an error without parsing the message), so I picked the doc-side fix.

All five Copilot threads resolved.

### Codecov patch coverage (59%)

The four flagged providers (BigQuery 27%, Postgres/Snowflake/Databricks 53%) are limited by Go's stdlib `database/sql` and the BigQuery SDK exposing `*sql.Rows` and `*bq.Client` as **concrete** types — they cannot be substituted via interface or instantiated outside the real driver, so the `rows.Next()` / `q.Run().LastStatus()` success branches are only reachable via integration tests.

This is the same coverage shape the rest of `providers/warehouse/*/provider.go` already has (BigQuery's project baseline is ~17%, Snowflake ~66%, Postgres similar). The project has historically not used `go-sqlmock` to work around concrete types.

What IS unit-tested per provider:
- empty / whitespace SQL boundary
- the wire format the implementation hands to the client (e.g. `EXPLAIN` prefix, `SET NOEXEC ON ... OFF` batch)
- error propagation wrapping

What IS integration-tested per provider (via build tags + run-integration-tests label):
- Postgres: testcontainers — `make test-postgres` covers valid SELECT, malformed SQL, empty input, quoted-identifier-as-keyword edge case
- BigQuery: real — `make test-bigquery` against `demo_olist_ecommerce.customers` covers valid SELECT, malformed SQL, missing table, empty input
- Redshift: real — `make test-redshift` against `default-workgroup` covers the same matrix
- MSSQL: testcontainers — `make test-mssql` covers the matrix PLUS a regression-pin assertion that INSERT through NOEXEC does not modify the table (so a future revert to PARSEONLY fails loudly)
- Snowflake / Databricks: unit-only — no test accounts available yet

I considered adding the four files to the codecov ignore list per the engineering convention, but that would hide coverage signal for the rest of each file (Query, ListTables, GetTableSchema, etc.). Leaving the codecov check red — it's signaling, not blocking — rather than mask the rest of the file.

If we get a Snowflake / Databricks test account later, the unit-only branches collapse to "covered by integration tests" too.
